Source/WebCore/page/Quirks.cpp

```diff
@@ -2087,19 +2081,28 @@ bool Quirks::needsChromeMediaControlsPseudoElement() const
 }
 
 #if PLATFORM(IOS_FAMILY)
-// Remove this once rdar://139478801 is resolved.
-bool Quirks::shouldSynthesizeTouchEventsAfterNonSyntheticClick(const Node& target) const
+// cbssports.com <rdar://139478801>.
+// docs.google.com <rdar://59402637>.
+bool Quirks::shouldSynthesizeTouchEventsAfterNonSyntheticClick(const Element& target) const
 {
     if (!needsQuirks())
         return false;
 
-    if (!m_quirksData.shouldSynthesizeTouchEventsAfterNonSyntheticClickQuirk)
-        m_quirksData.shouldSynthesizeTouchEventsAfterNonSyntheticClickQuirk = isDomain("cbssports.com"_s);
+    if (isCBSSports())
+        return target.nodeName() == "AVIA-BUTTON"_s;
 
-    if (!m_quirksData.shouldSynthesizeTouchEventsAfterNonSyntheticClickQuirk.value())
-        return false;
+    if (isGoogleDocs()) {
+        unsigned numberOfAncestorsToCheck = 3;
+        for (Ref ancestor : lineageOfType<HTMLElement>(target)) {
+            if (ancestor->hasClassName("docs-ml-promotion-action-container"_s))
+                return true;
 
-    return target.nodeName() == "AVIA-BUTTON"_s;
+            if (!--numberOfAncestorsToCheck)
+                break;
+        }
```

Comment:
I wonder if the iterator this returns is already compatible with std::ranges, in which case we could do something like:

```
constexpr unsigned numberOfAncestorsToCheck = 3;
for (Ref ancestor : lineageOfType<HTMLElement>(target) | std::ranges::views::take(numberOfAncestorsToCheck)) {
    if (ancestor->hasClassName("docs-ml-promotion-action-container"_s))
        return true;
}
```

Reply:
Oh, interesting — let me try that...